tests/mltoolbox_magic/local_predict_tests.py

```diff
@@ -0,0 +1,169 @@
+      }
+      model_dir = os.path.join(self._test_dir, dir_name)
+      builder = tf.saved_model.builder.SavedModelBuilder(model_dir)
+      builder.add_meta_graph_and_variables(
+          sess, [tf.saved_model.tag_constants.SERVING],
+          signature_def_map=signature_def_map)
+      builder.save(False)
+
+    return model_dir
+
+  def _create_test_data(self, embedding_images=False, missing_values=False, csv_data=False):
+    image_path1 = os.path.join(self._test_dir, 'img1.jpg')
+    image_path2 = os.path.join(self._test_dir, 'img2.jpg')
+    image_path3 = os.path.join(self._test_dir, 'img3.jpg')
+    Image.new('RGBA', size=(128, 128), color=(155, 211, 64)).save(image_path1, "JPEG")
+    Image.new('RGB', size=(64, 64), color=(111, 21, 86)).save(image_path2, "JPEG")
+    Image.new('RGBA', size=(16, 16), color=(255, 21, 1)).save(image_path3, "JPEG")
+
+    data = [
+        {'key': 1, 'num1': 4.1, 'text1': '32', 'img_url1': image_path1},
+        {'key': 2, 'num1': 5.2, 'text1': '18', 'img_url1': image_path2},
+        {'key': 5, 'num1': -12.0, 'text1': '22', 'img_url1': image_path3},
+    ]
+    if embedding_images:
+      for d in data:
+        with open(d['img_url1'], 'r') as f:
+          d['img_url1'] = base64.urlsafe_b64encode(f.read())
+
+    if missing_values:
+      del data[0]['num1']
+      del data[1]['img_url1']
+
+    csv_lines = []
+    if csv_data:
+      for d in data:
+        buf = six.StringIO()
+        writer = csv.DictWriter(buf, fieldnames=['key', 'num1', 'text1', 'img_url1'])
+        writer.writerow(d)
+        csv_lines.append(buf.getvalue().rstrip())
+      data = csv_lines
+
+    return data
+
+  def _validate_results(self, df, show_image):
+    expected_columns = set(['key', 'num1', 'text1', 'img_url1', 'var1', 'var2'])
+    if show_image:
+      expected_columns.add('img_url1_image')
+    self.assertEqual(expected_columns, set(df.columns))
+    self.assertEqual([1, 2, 5], df['key'].tolist())
+    self.assertEqual(3, len(df.index))
+
+  def test_predict(self):
+    """ Test prediction on a model which accepts CSV lines "int64,float32,text,image_url".
+    """
+
+    model_dir = self._create_model('model1')
+    headers = ['key', 'num1', 'text1', 'img_url1']
+
+    # Test data being list of dict and list of csvlines, with and without missing values,
+    # show or not show images.
+    for missing_values in [True, False]:
+      for csv_data in [True, False]:
+        for show_image in [True, False]:
+          test_data = self._create_test_data(False, missing_values, csv_data)
+          df = _local_predict.get_prediction_results(
+              model_dir, test_data, headers, ['img_url1'], show_image)
+          self._validate_results(df, show_image)
+
+    # Test data being dataframes, with and without missing values, and embedded images.
+    for missing_values in [True, False]:
+      test_data = self._create_test_data(True, missing_values, csv_data=False)
+      df_s = pd.DataFrame(test_data).fillna('')
```

Comment:
Can't. Otherwise %%ml predict will need to access stats files, which is not possible in online prediction.